source_pubsub/client.go

```diff
@@ -0,0 +1,520 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package source
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/conduitio-labs/conduit-connector-salesforce/source_pubsub/proto"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/linkedin/goavro/v2"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/metadata"
+	"gopkg.in/tomb.v2"
+)
+
+var (
+	// topic and subscription-related variables.
+
+	// gRPC server variables.
+	GRPCEndpoint    = "api.pubsub.salesforce.com:7443"
+	GRPCDialTimeout = 5 * time.Second
+	GRPCCallTimeout = 5 * time.Second
+)
+
+type PubSubClient struct {
+	accessToken string
+	instanceURL string
+
+	userID string
+	orgID  string
+
+	conn         *grpc.ClientConn
+	pubSubClient proto.PubSubClient
+	subClient    proto.PubSub_SubscribeClient
+	mu           *sync.Mutex
+
+	codecCache  map[string]*goavro.Codec
+	unionFields map[string]map[string]struct{}
+
+	buffer chan sdk.Record
+	caches chan []ConnectResponseEvent
+	ticker *time.Ticker
+
+	currreplayID []byte
+	tomb         *tomb.Tomb
+	topic        string
+}
+
+type ConnectResponseEvent struct {
+	Data     map[string]interface{}
+	eventID  string
+	replayID []byte
+}
+
+// Creates a new connection to the gRPC server and returns the wrapper struct.
+func NewGRPCClient(ctx context.Context, pollingPeriod time.Duration, config Config, sdkPos sdk.Position) (*PubSubClient, error) {
+	cx, cancelFn := context.WithTimeout(ctx, GRPCDialTimeout)
+	defer cancelFn()
+
+	sdk.Logger(cx).Debug().Msg("NewGRPCClient - Starting GRPC Client")
+
+	dialOpts := []grpc.DialOption{
+		grpc.WithBlock(),
+	}
+
+	if GRPCEndpoint == "localhost:7011" {
+		dialOpts = append(dialOpts, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	} else {
+		certs := getCerts()
+		creds := credentials.NewClientTLSFromCert(certs, "")
+		dialOpts = append(dialOpts, grpc.WithTransportCredentials(creds))
+	}
+
+	conn, err := grpc.DialContext(cx, GRPCEndpoint, dialOpts...)
+	if err != nil {
+		return nil, fmt.Errorf("error on dialing grpc - %s", err)
+	}
+
+	sdk.Logger(cx).Debug().Msg("NewGRPCClient - Initialize Pubsub")
+
+	pubSub := &PubSubClient{
+		conn:         conn,
+		pubSubClient: proto.NewPubSubClient(conn),
+		codecCache:   make(map[string]*goavro.Codec),
+		unionFields:  make(map[string]map[string]struct{}),
+		buffer:       make(chan sdk.Record, 1),
+		caches:       make(chan []ConnectResponseEvent),
+		mu:           &sync.Mutex{},
+		ticker:       time.NewTicker(pollingPeriod),
+		tomb:         &tomb.Tomb{},
```

Comment:
You want to pass this context to the tomb, in case it gets cancelled.

```suggestion
		tomb:         tomb.WithContext(ctx),
```